sync/sync.py

```diff
@@ -58,76 +55,77 @@
+        for line in fileinput.input(file_in_folder, inplace=1):
             # add a line of text to the payload
             # transform_links mutates text and set the lines provided
-            transform_links(line, url, link_prefix)
+            print(transform_line(line, folder, base_path, base_url))
 
     with Pool() as pool:
-        pool.imap_unordered(process_file, dest_files)
-
-
-def github_link(url, link):
-    """ given a github raw link convert it to the main github link """
-    return f'{url.replace("raw", "tree", 1)}/{link}'
+        pool.imap_unordered(process_file, files_in_folder)
 
 
-def sanitize_text(link_prefix, text):
-    """ santize every line of text to exclude relative
-    links and to turn markdown file URL's to html """
-    old_line = text.rstrip()
-    new_line = re.sub(LINKS_RE, r'[\1](' + link_prefix + r'\2\3)', old_line)
-    return new_line, old_line == new_line
-
-
-def is_url(url):
-    """ check if it is a valid url """
-    try:
-        urlopen(url).read()
-    except (HTTPError, URLError):
-        return True
-    except ValueError:
-        return False
-
-    return True
-
-
-def is_ref(url):
+def transform_line(line, base_path, rewrite_path, rewrite_url):
+    """ transform all the links in one line """
+    line = line.rstrip()
+    links = get_links(line)
+    # If there are links in this line we may need to fix them
+    for link in links:
+        # link contains the text and href
+        href =link.get("href")
+        href_mod = transform_link(href, base_path, rewrite_path, rewrite_url)
+        line = line.replace(href, href_mod)
+    return line
+
+
+def transform_link(link, base_path, rewrite_path, rewrite_url):
+    """ Transform hrefs to be valid URLs on the web-site
+
+    Absolute URLs are not changed (they may be external)
+    Fragments are relative to the page and do not need changes
+    Path only links should point to a file synced to the website
+    but sometimes the file may be missing (if it's not in the sync
+    configuration), so we follow this approach:
+    - prefix with base_path and check for the file locally
+    - if not found, prefix with base_url instead
+
+    Note that urlparse treats URLs without scheme like path only
+    URLs, so 'github.com' will be rewritten to base_url/github.com
+    """
+
+    # ignore empty links
+    if not link:
+        return link
+    # urlparse returns a named tuple
+    parsed = urlparse(link)
+    if is_absolute_url(parsed) or is_fragment(parsed):
+        return link
+    path = os.path.normpath(parsed.path)
+    if os.path.isfile(os.path.join(base_path, path)):
+        filename, ext = os.path.splitext(path)
+        # md files links are in the format .../[md filename]/
+        if ext == '.md':
+            path = filename + '/'
+        return urlunparse(parsed._replace(path="/".join([rewrite_path, path])))
+    # when not found on disk, append to the base_url
+    return urljoin(rewrite_url, link)
```

Comment:
This transform links does not transform links correctly.

Reply:
Uhm, why not? Do you have an example?